scripts/constituents.py

```diff
@@ -484,124 +468,171 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
         # end for
         cap.comment("Local variables", 2)
         spc = ' '*37
-        cap.write("integer{} :: num_suite_consts".format(spc), 2)
-        cap.write("integer{} :: num_consts".format(spc), 2)
-        cap.write("integer{} :: index".format(spc), 2)
-        cap.write("integer{} :: field_ind".format(spc), 2)
+        cap.write(f"integer{spc} :: num_suite_consts", 2)
+        cap.write(f"integer{spc} :: num_consts", 2)
+        cap.write(f"integer{spc} :: num_dyn_consts", 2)
+        cap.write(f"integer{spc} :: index, index_start", 2)
+        cap.write(f"integer{spc} :: field_ind", 2)
         cap.write(f"type({CONST_PROP_TYPE}), pointer :: const_prop", 2)
+        # Declare dynamic constituent properties variables
+        for idx, scheme in enumerate(sorted(dyn_const_dict)):
+            cap.comment(f"dynamic constituent props variable for {scheme}", 2)
+            cap.write(f"type({CONST_PROP_TYPE}), allocatable :: dyn_const_prop_{idx}(:)", 2)
+        # end for
         cap.blank_line()
-        cap.write("{} = 0".format(herrcode), 2)
+        cap.write(f"{herrcode} = 0", 2)
         cap.write("num_consts = size(host_constituents, 1)", 2)
+        cap.write("num_dyn_consts = 0", 2)
         for suite in suite_list:
             const_dict = suite.constituent_dictionary()
             funcname = const_dict.num_consts_funcname()
             cap.comment(f"Number of suite constants for {suite.name}", 2)
             errvar_str = ConstituentVarDict.__errcode_callstr(herrcode,
                                                               herrmsg, suite)
             cap.write(f"num_suite_consts = {funcname}({errvar_str})", 2)
+            cap.write(f"if ({herrcode} /= 0) then", 2)
+            cap.write("return", 3)
+            cap.write("end if", 2)
             cap.write("num_consts = num_consts + num_suite_consts", 2)
         # end for
-        cap.write("if ({} == 0) then".format(herrcode), 2)
-        cap.comment("Initialize constituent data and field object", 3)
-        stmt = "call {}%initialize_table(num_consts)"
-        cap.write(stmt.format(const_obj_name), 3)
+        # Check for dynamic constituent routines
+        if len(dyn_const_dict) > 0:
+            cap.comment("Add in dynamic constituents", 2)
+            for idx, scheme in enumerate(sorted(dyn_const_dict)):
+                cap.write(f"call {dyn_const_dict[scheme]}(dyn_const_prop_{idx}, {obj_err_callstr})", 2)
+                cap.write(f"if ({herrcode} /= 0) then", 2)
+                cap.write("return", 3)
+                cap.write("end if", 2)
+                cap.write(f"num_dyn_consts = num_dyn_consts + size(dyn_const_prop_{idx})", 2)
+            # end for
+            cap.write("num_consts = num_consts + num_dyn_consts", 2)
+            cap.comment("Pack dynamic_constituents array", 2)
+            cap.write(f"allocate(dynamic_constituents(num_dyn_consts), stat={herrcode})", 2)
```

Comment:
I do not understand the need for `dynamic_constituents`. Why not just copy out of the arrays returned by the schemes?

Reply:
I remember now (after trying to remove it) why I passed in the dynamic_constituents array from the host model. It's because the dyn_const_prop_x arrays go out of scope when we leave the register_constituents routine. Having the constituent properties array point to dynamic constituents that are declared at the host level keeps them in scope.

That said, there certainly could be a better way to do all of this!

Reply:
It seems inconvenient to ask the host model to do memory management for the framework. There could be a module variable for each suite's dynamic constituents.

Speaking of multiple suites, does this routine work if more than one suite is built in? For the solution above, I would expect one (allocateable or pointer) module variable per suite.

Reply:
Updated to use a single dynamic constituent array at the host cap level because it seemed easier.

To answer your other question: it does work if there's more than one suite - there are just more dynamic constituent routines called within register.

Reply:
Now have this at the top of <host>_ccpp_cap.F90:

`type(ccpp_constituent_properties_t), allocatable, target :: test_host_dynamic_constituents(:)`

And we're no longer passing in dynamic_constituents from the host and instead allocating/packing the test_host_dynamic_constituents array.

Reply:
I guess my concern is for a model such as UFS, there will be two suites but they do not have the same data. I don't know if they currently have non-advected constitutents but imagine that the physics has some of those. The "fast physics" suite that runs in the dycore only has the advected constituents and operates on different data so I think we will need multiple dynamic constituents arrays.

On the other hand, models such as the UFS will probably need to present as two host models which would lead to two host caps. Given the different "host data", I think this makes the most sense but we need to figure out how we are going to package and process the combined suites. Someone from NOAA correct me if I'm off base here.

In any case, we need to decide what types of arrangements will be supported to see if your current design will work in all those cases.

Reply:
I _think_ I understand, and based on that limited understanding, I would think that, yes, models like UFS will need two host caps (we'll need two constituent objects, not just two dynamic constituent arrays).

Reply:
opened related issue #561

Reply:
> I _think_ I understand, and based on that limited understanding, I would think that, yes, models like UFS will need two host caps (we'll need two constituent objects, not just two dynamic constituent arrays).

Maybe open an issue on this? This is sort of a combo framework <==> host model issue so not sure where it goes.